### PR TITLE
WIP: Native downloader

### PR DESCRIPTION
**Pull request by Chasbob**
Don't touch, just pray.

### CODE REVIEW

File: bhamcal/frame.py [hunk 1]
```diff
@@ -129,3 +177,7 @@ def FIREFOX(headless=True):
         options.add_argument('--headless')
     driver = webdriver.Firefox(options=options)
     return driver
+
+if __name__ == "__main__":
+    frame = NativeFrame()
+    frame.fetch("cxd738","LEFTIES#dacquoise#RACIALIZED")
```
Comment (Chasbob) [inline bhamcal/frame.py]: Oops....
Has been fixed.

**Comment by jedevc** [comment]
Eyyyy awesome :tada: :tada: I'd almost given up on this working :cry:

File: bhamcal/frame.py (hunk 1 above)
Comment (jedevc) [reply bhamcal/frame.py]: Congrats :)

File: bhamcal/frame.py [hunk 2]
```diff
@@ -34,30 +34,59 @@ def fetch(self, username, password):
...
-        pprint(filled)
-
-        resp = session.post(urljoin(resp.url, form['action']), data=filled, headers=NativeFrame.HEADERS)
-        with open('dump.2.html', 'w') as dump:
-            dump.write(resp.text)
+        # Find all available module codes
+        select_modules = soup.find('select', attrs={'name': 'dlObject'})
+        modules = [x['value'] for x in select_modules.findChildren()]
+
+        headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:72.0) Gecko/20100101 Firefox/72.0", "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8", "Accept-Language": "en-GB,en;q=0.5", "Accept-Encoding": "gzip, deflate", "Content-Type": "application/x-www-form-urlencoded", "Origin": "https://onlinetimetables.bham.ac.uk", "DNT": "1", "Connection": "close", "Referer": "https://onlinetimetables.bham.ac.uk/timetable/current_academic_year_2/default.aspx", "Upgrade-Insecure-Requests": "1"}
```
Comment (jedevc) [inline bhamcal/frame.py]: Why the headers inline here when `NativeFrame.HEADERS` is used above?

File: bhamcal/frame.py [hunk 3]
```diff
@@ -34,30 +34,59 @@ def fetch(self, username, password):
...
+                        # Select weeks
+                        "lbWeeks": soup.find('option', string='*All Term Time')['value'],
+                        # Select days of the week
+                        "lbDays": soup.find('option', string='All Week')['value'],
+                        # Select hours of the day to view
+                        "dlPeriod": soup.find('option', string='All Day (08:00 - 22:00)')['value'],
+                        # Formate to output the timetable in
+                        "dlType": soup.find('option', string='List Timetable (with calendar dates)')['value'],
+                        # "dlType": "TextSpreadsheet;swsurl;SWSCUST Object TextSpreadsheet MDS",
+                        "bGetTimetable": "View Timetable"
```
Comment (jedevc) [inline bhamcal/frame.py:84]: There's no comment for this item, but comments for all the other items.

File: bhamcal/frame.py [hunk 4]
```diff
@@ -34,30 +34,59 @@ def fetch(self, username, password):
...
+                        "dlObject": modules,
+                        # Select weeks
+                        "lbWeeks": soup.find('option', string='*All Term Time')['value'],
+                        # Select days of the week
+                        "lbDays": soup.find('option', string='All Week')['value'],
+                        # Select hours of the day to view
+                        "dlPeriod": soup.find('option', string='All Day (08:00 - 22:00)')['value'],
+                        # Formate to output the timetable in
+                        "dlType": soup.find('option', string='List Timetable (with calendar dates)')['value'],
+                        # "dlType": "TextSpreadsheet;swsurl;SWSCUST Object TextSpreadsheet MDS",
```
Comment (jedevc) [inline bhamcal/frame.py]: This can be removed?

File: bhamcal/frame.py [hunk 5]
```diff
@@ -129,3 +158,7 @@ def FIREFOX(headless=True):
         options.add_argument('--headless')
     driver = webdriver.Firefox(options=options)
     return driver
+
+if __name__ == "__main__":
```
Comment (jedevc) [inline bhamcal/frame.py]: Can this main be removed from here as it's only for testing?

File: bhamcal/frame.py [hunk 6]
```diff
@@ -34,30 +34,59 @@ def fetch(self, username, password):
...
-
-        resp = session.post(urljoin(resp.url, form['action']), data=filled, headers=NativeFrame.HEADERS)
-        with open('dump.2.html', 'w') as dump:
-            dump.write(resp.text)
+        # Find all available module codes
+        select_modules = soup.find('select', attrs={'name': 'dlObject'})
+        modules = [x['value'] for x in select_modules.findChildren()]
+
+        headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:72.0) Gecko/20100101 Firefox/72.0", "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8", "Accept-Language": "en-GB,en;q=0.5", "Accept-Encoding": "gzip, deflate", "Content-Type": "application/x-www-form-urlencoded", "Origin": "https://onlinetimetables.bham.ac.uk", "DNT": "1", "Connection": "close", "Referer": "https://onlinetimetables.bham.ac.uk/timetable/current_academic_year_2/default.aspx", "Upgrade-Insecure-Requests": "1"}
+        formdata = {
```
Comment (jedevc) [inline bhamcal/frame.py:59]: Indentation should be only 4 spaces.

**Review by jedevc — COMMENTED** [review]
Just nitpicks before merging.

**Comment by jedevc** [comment]
Merged manually!

**Pull request closed without merging**